cpp/src/cider/common/Arena.h

```diff
@@ -40,35 +54,137 @@ namespace cider::hashtable {
  */
 class Arena : private boost::noncopyable {
  private:
-  static size_t roundUpToPageSize(size_t s, size_t page_size) { return 0; }
+  static constexpr size_t pad_right = PADDING_FOR_SIMD - 1;
+
+  // Contiguous MemoryChunk of memory and pointer to free space inside it. Member of
+  // single-linked list.
+  struct alignas(16) MemoryChunk : private CiderDefaultAllocator {
+    char* begin;
+    char* pos;
+    char* end;  // does not include padding.
+
+    MemoryChunk* prev;
+
+    MemoryChunk(size_t size_, MemoryChunk* prev_) {
+      begin = reinterpret_cast<char*>(CiderDefaultAllocator::allocate(size_));
+      pos = begin;
+      end = begin + size_ - pad_right;
+      prev = prev_;
+    }
+
+    ~MemoryChunk() {
+      CiderDefaultAllocator::deallocate(reinterpret_cast<int8_t*>(begin), size());
+      delete prev;
+    }
+
+    size_t size() const { return end + pad_right - begin; }
```

Comment:
why not use a class member to track size?

Reply:
reduce extra memory cost